server/log_rpc_server.go

```diff
@@ -406,26 +395,75 @@ func (t *TrillianLogRPCServer) GetConsistencyProof(ctx context.Context, req *tri
 func (t *TrillianLogRPCServer) GetLatestSignedLogRoot(ctx context.Context, req *trillian.GetLatestSignedLogRootRequest) (*trillian.GetLatestSignedLogRootResponse, error) {
 	ctx, span := spanFor(ctx, "GetLatestSignedLogRoot")
 	defer span.End()
-	tree, ctx, err := t.getTreeAndContext(ctx, req.LogId, optsLogRead)
+	tree, hasher, err := t.getTreeAndHasher(ctx, req.LogId, optsLogRead)
 	if err != nil {
 		return nil, err
 	}
+	ctx = trees.NewContext(ctx, tree)
 	tx, err := t.registry.LogStorage.SnapshotForTree(ctx, tree)
 	if err != nil {
 		return nil, err
 	}
 	defer tx.Close()
 
-	signedRoot, err := tx.LatestSignedLogRoot(ctx)
+	slr, err := tx.LatestSignedLogRoot(ctx)
 	if err != nil {
 		return nil, err
 	}
 
+	var root types.LogRootV1
+	if err := root.UnmarshalBinary(slr.LogRoot); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not read current log root: %v", err)
+	}
+
+	r := &trillian.GetLatestSignedLogRootResponse{SignedLogRoot: &slr}
+
+	if req.FirstTreeSize == 0 {
+		// no need to get consistency proof in this case
+		if err := t.commitAndLog(ctx, req.LogId, tx, "GetLatestSignedLogRoot"); err != nil {
+			return nil, err
+		}
+		return r, nil
+	}
+
+	reqProof := &trillian.GetConsistencyProofRequest{
+		LogId:          req.LogId,
+		FirstTreeSize:  int64(req.FirstTreeSize),
+		SecondTreeSize: int64(root.TreeSize),
+	}
+	if err := validateGetConsistencyProofRequest(reqProof); err != nil {
+		return nil, err
```

Comment:
Hmm, this will give somewhat different error behaviour for server skew than the existing code.

 - If a `GetLatestSLR` request hits (say) a server that's way behind other servers (and so isn't yet aware of tree size `FirstTreeSize`), this code will emit `InvalidArgument` and no response.
 - For other entrypoints (e.g. `GetConsistencyProof`, `GetInclusionProofByHash`), hitting a behind-server tends to result in an OK response that has the known SLR but other fields missing.

(Sadly, the existing skew behaviour is mostly undocumented&hellip;)

Reply:
I've opened an issue for further discussion: #1472

Reply:
It seems back to front to change API behaviour and then afterwards raise an issue to discuss changing API behaviour.

Reply:
Sorry about that.  Just trying to make forward progress.  
This seemed OK to me to merge since it didn't disrupt any existing API behaviors.